### ISSUE TITLE
the inference speed of fp16 is almost the same as fp32

### ISSUE THREAD

**Issue by wolfshow** [pytorch/fairseq#714]
In the training, fp16 runs much faster, but during inference, it is almost the same as fp32. Both training and testing use max-tokens of 4,000 on a V100 GPU. Why?

**Comment by wolfshow** [comment]
According to https://code.fb.com/ai-research/scaling-neural-machine-translation-to-bigger-data-sets-with-faster-training-and-inference/, Simply changing from Float 32 to Float 16 improved speed by 40 percent.

**Comment by frankang** [comment]
In the article it mentions

> In particular, we implemented clever caching, or removing finished sentences from the computation and batching by the number of words instead of sentences

I don't know what clever caching is, but they might have increased the max-tokens for the fp16 model as it is able to hold more sentences in GPU memory.

You may further analyze the function call time by using some profiling tools such as pyflame in python

**Comment by alexeib** [comment]
the "clever caching" is not fp16 specific. 

frankang is right, to take full advantage of fp16 you should increase max-tokens. what model are you measuring speed with?

**Comment by wolfshow** [comment]
> the "clever caching" is not fp16 specific.
> 
> frankang is right, to take full advantage of fp16 you should increase max-tokens. what model are you measuring speed with?

unfortunately, we cannot increase max-tokens. We are testing the inference speed for fixed batch size from fp32 to fp16.

**Comment by alexeib** [comment]
in this case you are unlikely to see any gains
